### ISSUE TITLE
setuid root executables lose sticky bit

### ISSUE THREAD

**Issue by d** [concourse/concourse#802]
# Bug Report

* Concourse version: 2.5.0
* Deployment type (BOSH/Docker/binary): BOSH, but should be applicable to all
* Infrastructure/IaaS: AWS, but really any

## Steps to reproduce

1. Create a task yaml, which:
    1. Uses `image_resource` instead of `image`
    1. The image specified as such contains setuid root executable, e.g. `/bin/ping`
    1. Runs as a non-root user, and call the setuid-root executable

1. `fly execute` said task

### Expected result
This should work

### Actual result
The setuid-root executable from the Docker image will lose its sticky bit, and hence this would result in insufficient permission error

**Comment by illegalprime** [comment]
I'm also experiencing this

**Comment by miromode** [comment]
Same here, very much a problem with `/usr/bin/sudo`

**Comment by ashb** [comment]
I just had this by using a task with `image` too, where the image was build in a previous resource.

I verified that the binary (`/usr/bin/sudo` in the image pushed to the (private) registry does have the sticky bit, and it does.

**Comment by jbraswell** [comment]
This is a major issue for me as well. I'm using v3.0.1. Would love to see a fix for this.

**Comment by kelvich** [comment]
+1

**Comment by craigfurman** [comment]
We've been seeing this too, but only after upgrading from 3.0.1 to 3.2.1. We are using the btrfs storage driver

**Comment by vito** [comment]
This is more pronounced with 3.1; I'll prioritize it somewhere soon.

I think we can just change https://github.com/concourse/baggageclaim/blob/c12e0c4e85ad48d068aefb761e65ca704b2a9dcd/uidgid/translator.go#L38 to check if setuid/setgid are set before the chown, and if so, restore it after the chown. If someone wants to PR that with integration test coverage in baggageclaim, that'd be super helpful!

**Comment by mhuangpivotal** [comment]
@d, we tried recreating the issues with the steps. Can you provide more explicit steps to recreate a docker image with the `/bin/ping` with the sticky bit? We were unable to recreate the issue.

**Comment by topherbullock** [comment]
Repro'd with 

`Dockerfile`:
```
FROM ubuntu
ADD hi.sh /root/hi.sh
RUN chmod u+s ~/hi.sh
```

`task.yml`:
```
platform: linux
image_resource:
  type: docker-image
  source: {repository: concourse/setuid-test }
run:
  path: ls
  args: ["-l","/root/"]
```

When we run `docker run <image> ls -l /root` we see the following:
`-rwsr-xr-x 1 root root 39 Aug  4 18:20 hi.sh`

When we `fly excute -c ./task.yml` we see the following (after the image pulls and such):
`-rwxr-xr-x 1 root root 39 Aug  4 18:20 hi.sh`

**Comment by vito** [comment]
Verified by hijacking into a task and checking that `ping` is setuid and able to run as non-root.

**Comment by thardeck** [comment]
Is this supposed to be fixed in 3.4.0 or in a later 3.4.x version? We still experience this or a similar issue in a freshly installed 3.4.0.

We have downgraded because of this issue to 3.0.1 some time ago which works fine.

This is the error:
```
/mnt/stemcells/null/null/ubuntu/build/build/stages/base_debootstrap/config.sh /mnt/stemcells/null/null/ubuntu/build/build/etc/settings.bash 2>&1
sudo: /usr/bin/sudo must be owned by uid 0 and have the setuid bit set
```

**Comment by vito** [comment]
It is in 3.4. Can you ls -al the file so that we can see the permissions?

**Comment by jtarchie** [comment]
This is a potentially a different error. The `sudo` command is saying that it is not owned by the root user. If a `chown` happened on it, even with this change we made, it will still not work.

**Comment by thardeck** [comment]
```
ls -al /usr/bin/sudo
-rwxrwxrwx 1 root root 155008 Oct 14  2016 /usr/bin/sudo
```

**Comment by jtarchie** [comment]
@thardeck, do you know what filesystem you are using for baggageclaim? overlay, btrfs, naive?

**Comment by thardeck** [comment]
``` 
- name: baggageclaim
    release: concourse
    properties: {driver: btrfs}
```

**Comment by tjvman** [comment]
We encountered this issue in `3.4.1` in one of our containers that runs as privileged, regardless of whether our worker's baggageclaim was using btrfs or overlayfs. 

Running `stat $(which sudo)` after hijacking in shows 0777 permissions. Manually running `chmod u+s $(which sudo)` in our task script fixed the issue. 

When we `docker run` that image locally, the same stat command shows the setuid bit active.

**Comment by thardeck** [comment]
@tjvman 
Are you planning to add a general fix for the issue to Concourse? Should I create a new issue?

**Comment by vito** [comment]
@thardeck Yeah, please open a new issue.

I'm now thinking we may have fixed this in one place, but in other scenarios, e.g. when we need to transfer bits between workers, `tar` or something is stripping the bits.

Having a minimal repro for this would help, as we at least saw that one case was fixed. Can we try restricting it to one worker, with `overlay` (default driver), for example?

**Comment by thardeck** [comment]
@vito I have create a new issue under #1569.
